### PR TITLE
[Streaming] Add configuration with owner config.

### PR DESCRIPTION
**Pull request by BalaBalaYi**
## Why are these changes needed?
Add basic configuration implementations in streaming runtime.

The following PRs like 'graph', 'resource' or 'scheduler',  will extend the configuration for each part based on the current approach.

## Related issue number
None

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/StreamingConfig.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+import java.io.Serializable;
+import java.util.Map;
+
+/**
+ * Streaming config including general, master and worker part.
+ */
+public class StreamingConfig implements Serializable {
+
+  public StreamingMasterConfig masterConfig;
```
Comment (ffbin) [inline streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/StreamingConfig.java]: setter and getter?

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/StreamingConfig.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  public StreamingWorkerConfig workerConfigTemplate;
+
+  public StreamingConfig(Map<String, String> conf) {
```
Comment (ffbin) [inline streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/StreamingConfig.java]: const Map<String, String> conf ?

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/StreamingMasterConfig.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    // configMap.putAll(masterConfig2Map());
+  }
+
+  @Override
+  public String toString() {
+    return configMap.toString();
+  }
+
+  /*
+  public Map<String, String> masterConfig2Map() {
```
Comment (ffbin) [inline streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/StreamingMasterConfig.java]: remove it

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/StreamingWorkerConfig.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  private static final Logger LOGGER = LoggerFactory.getLogger(StreamingWorkerConfig.class);
+
+  public StreamingWorkerConfig(Map<String, String> conf) {
+    super(conf);
+
+    // configMap.putAll(workerConfig2Map());
+  }
+
+  /*
+  public Map<String, String> workerConfig2Map() {
```
Comment (ffbin) [inline streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/StreamingWorkerConfig.java]: ditto

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/global/CommonConfig.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
+package org.ray.streaming.runtime.config.global;
+
+import org.ray.streaming.runtime.config.Config;
+
+/**
+ * Job common config.
+ */
+public interface CommonConfig extends Config {
+
+  String JOB_ID = "streaming.job.id";
```
Comment (ffbin) [inline streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/global/CommonConfig.java]: add some comments

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/StreamingConfig.java (hunk 1 above)
Comment (BalaBalaYi) [reply streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/StreamingConfig.java]: StreamingConfig will be hold by runtime context with 'private'. No need for setter and getter here.

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/StreamingMasterConfig.java (hunk 3 above)
Comment (BalaBalaYi) [reply streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/StreamingMasterConfig.java]: Removed.

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/global/CommonConfig.java (hunk 5 above)
Comment (BalaBalaYi) [reply streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/global/CommonConfig.java]: Maybe add '@Description' on config prop?

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/StreamingConfig.java (hunk 2 above)
Comment (BalaBalaYi) [reply streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/StreamingConfig.java]: fixed

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/StreamingConfig.java [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+import java.io.Serializable;
+import java.util.Map;
+
+/**
+ * Streaming config including general, master and worker part.
+ */
+public class StreamingConfig implements Serializable {
+
+  public StreamingMasterConfig masterConfig;
+  public StreamingWorkerConfig workerConfigTemplate;
```
Comment (datayjz) [inline streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/StreamingConfig.java]: Can it be named `workConfig`?

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/StreamingMasterConfig.java [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+import java.util.Map;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Streaming job master config.
+ */
+public class StreamingMasterConfig extends StreamingGlobalConfig {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(StreamingMasterConfig.class);
```
Comment (datayjz) [inline streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/StreamingMasterConfig.java]: Global unified `LOG` variable name.

**Review by JingGe — CHANGES_REQUESTED** [review]

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/StreamingConfig.java (hunk 6 above)
Comment (BalaBalaYi) [reply streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/StreamingConfig.java]: Use as template for now.

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/util/TestHelper.java [hunk 8]
```diff
@@ -0,0 +1,18 @@
+package org.ray.streaming.runtime.util;
+
+public class TestHelper {
```
[the remark this answers is not in the document]
Comment (BalaBalaYi) [reply streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/util/TestHelper.java]: It is used for testing but not only used in test case. So keep it.

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/StreamingGlobalConfig.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+  @Override
+  public String toString() {
+    return configMap.toString();
+  }
+
+  public void globalConfig2Map() {
+    try {
+      configMap.putAll(config2Map(this.commonConfig));
+    } catch (Exception e) {
+      LOG.error("Global config to map occur error.", e);
```
Comment (raulchen) [inline streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/StreamingGlobalConfig.java]:
```suggestion
      LOG.error("Couldn't convert global config to a map.", e);
```

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/util/TestHelper.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (BalaBalaYi) [reply streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/util/TestHelper.java]: > where will this class be used, except test? And what is `UT_PATTERN`?

To increase coverage of ut in some cases. Or just need some implementations for ut only.
e.g.
```
if (is ut_pattern) {
   mock xxx ...
}
```

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/StreamingGlobalConfig.java (hunk 9 above)
Comment (BalaBalaYi) [reply streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/StreamingGlobalConfig.java]: Fixed.

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/StreamingGlobalConfig.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    commonConfig = ConfigFactory.create(CommonConfig.class, conf);
+    globalConfig2Map();
+  }
+
+  @Override
+  public String toString() {
+    return configMap.toString();
+  }
+
+  public void globalConfig2Map() {
```
[the remark this answers is not in the document]
Comment (BalaBalaYi) [reply streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/StreamingGlobalConfig.java]: globalConfig2Map() -> private
config2Map() -> protected

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/StreamingGlobalConfig.java [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+/**
+ * Streaming general config. May used by both JobMaster and JobWorker.
+ */
+public class StreamingGlobalConfig implements Serializable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(StreamingGlobalConfig.class);
+
+  public CommonConfig commonConfig;
+
+  public Map<String, String> configMap;
```
[the remark this answers is not in the document]
Comment (BalaBalaYi) [reply streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/StreamingGlobalConfig.java]: Made it final.

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/util/TestHelper.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (JingGe) [reply streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/util/TestHelper.java]: > It is used for testing but not only used in test case. So keep it.

Then a different name should be used.

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/global/CommonConfig.java [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+  @Key(value = JOB_NAME)
+  String jobName();
+
+  /**
+   * Default encoding type. e.g. UTF-8, GBK
+   * @return Encoding type with string type.
+   */
+  @DefaultValue(value = "UTF-8")
+  @Key(value = FILE_ENCODING)
+  String fileEncoding();
```
[the remark this answers is not in the document]
Comment (BalaBalaYi) [reply streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/global/CommonConfig.java]: There might be some cases that need to write content into temp files during streaming runtime.
So use UTF-8 as a default type and similar references are consistent in every streaming job.

File: streaming/java/streaming-runtime/src/test/java/org/ray/streaming/runtime/config/ConfigTest.java [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+  @org.testng.annotations.BeforeMethod
+  public void testBegin(Method method) {
+    LOG.warn(">>>>>>>>>>>>>>>>>>>> Test case: " + method.getName() + " begin >>>>>>>>>>>>>>>>>>");
+  }
+
+  @org.testng.annotations.AfterMethod
+  public void testEnd(Method method) {
+    LOG.warn(">>>>>>>>>>>>>>>>>>>> Test case: " + method.getName() + " end >>>>>>>>>>>>>>>>>>");
+  }
```
[the remark this answers is not in the document]
Comment (BalaBalaYi) [reply streaming/java/streaming-runtime/src/test/java/org/ray/streaming/runtime/config/ConfigTest.java]: Add ```BaseUnitTest``` as super.

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/util/TestHelper.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (BalaBalaYi) [reply streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/util/TestHelper.java]: Rename to ```TestUtil``` or ```MockUtil```?  Any better suggestion?
@JingGe @raulchen

[the remark this answers is not in the document]
Comment (JingGe) [reply streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/util/TestHelper.java]: Both TestUtil and MockUtil have the intent to use it ONLY for test. Please explain the usage of this class in the first place.

[the remark this answers is not in the document]
Comment (BalaBalaYi) [reply streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/util/TestHelper.java]: Move ```TestHelper``` to test folder.

File: streaming/java/streaming-runtime/src/test/java/org/ray/streaming/runtime/BaseUnitTest.java [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+    TestHelper.setUTPattern();
+  }
+
+  @AfterClass
+  public void tearDown() {
+    TestHelper.clearUTPattern();
+  }
+
+  @BeforeMethod
+  public void testBegin(Method method) {
```
Comment (ashione) [inline streaming/java/streaming-runtime/src/test/java/org/ray/streaming/runtime/BaseUnitTest.java]: Shall we always run testBegin and testEnd?

Comment (BalaBalaYi) [reply streaming/java/streaming-runtime/src/test/java/org/ray/streaming/runtime/BaseUnitTest.java]: Not essential. But adding this output is really good for problems positioning when debug unit tests' logs on server.

File: streaming/java/streaming-runtime/src/test/java/org/ray/streaming/runtime/BaseUnitTest.java [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+    LOG.info(">>>>>>>>>>>>>>>>>>>> Test case: " + method.getName() + " begin >>>>>>>>>>>>>>>>>>>>");
```
Comment (raulchen) [inline streaming/java/streaming-runtime/src/test/java/org/ray/streaming/runtime/BaseUnitTest.java]:
```suggestion
    LOG.info(">>>>>>>>>>>>>>>>>>>> Test case: " + method.getName() + " began >>>>>>>>>>>>>>>>>>>>");
```

File: streaming/java/streaming-runtime/src/test/java/org/ray/streaming/runtime/BaseUnitTest.java [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+  }
+
+  @AfterMethod
+  public void testEnd(Method method) {
+    LOG.info(">>>>>>>>>>>>>>>>>>>> Test case: " + method.getName() + " end >>>>>>>>>>>>>>>>>>");
```
Comment (raulchen) [inline streaming/java/streaming-runtime/src/test/java/org/ray/streaming/runtime/BaseUnitTest.java]:
```suggestion
    LOG.info(">>>>>>>>>>>>>>>>>>>> Test case: " + method.getName() + " ended >>>>>>>>>>>>>>>>>>");
```

File: streaming/java/streaming-runtime/src/test/java/org/ray/streaming/runtime/TestHelper.java [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+
+  public static void setUTPattern() {
+    System.setProperty(UT_PATTERN, "");
+  }
+
+  public static void clearUTPattern() {
+    System.clearProperty(UT_PATTERN);
+  }
+
+  public static boolean isUTPattern() {
```
Comment (raulchen) [inline streaming/java/streaming-runtime/src/test/java/org/ray/streaming/runtime/TestHelper.java]: Do you want to use this to tell if we are currently running tests?
What about just `isTest`? `ut pattern` is unclear.

Also, we don't need to set properties here. Just use a boolean flag.

**Review by raulchen — APPROVED** [review]

File: streaming/java/streaming-runtime/src/test/java/org/ray/streaming/runtime/TestHelper.java (hunk 17 above)
Comment (BalaBalaYi) [reply streaming/java/streaming-runtime/src/test/java/org/ray/streaming/runtime/TestHelper.java]: Fixed.

**Review by JingGe — APPROVED** [review]

**Pull request merged**